src/gf.c

```diff
@@ -681,26 +682,33 @@ static jl_lambda_info_t *cache_method(jl_methtable_t *mt, union jl_typemap_t *ca
     return newmeth;
 }
 
-static jl_lambda_info_t *jl_mt_assoc_by_type(jl_methtable_t *mt, jl_datatype_t *tt, int cache, int inexact)
+static jl_lambda_info_t *jl_mt_assoc_by_type(jl_methtable_t *mt, jl_datatype_t *tt, int cache, int inexact, jl_typemap_entry_t **entry)
 {
-    jl_typemap_entry_t *m = NULL;
+    *entry = NULL;
     jl_svec_t *env = jl_emptysvec;
     jl_method_t *func = NULL;
     jl_tupletype_t *sig = NULL;
-    JL_GC_PUSH4(&env, &m, &func, &sig);
+    JL_GC_PUSH4(&env, entry, &func, &sig);
 
-    m = jl_typemap_assoc_by_type(mt->defs, tt, &env, inexact, 1, 0);
-    if (m == NULL) {
+    *entry = jl_typemap_assoc_by_type(mt->defs, tt, &env, inexact, 1, 0);
+    if (*entry == NULL) {
         JL_GC_POP();
         return NULL;
     }
 
-    sig = join_tsig(tt, m->sig);
+    jl_method_t *m = (*entry)->func.method;
+    sig = join_tsig(tt, (*entry)->sig);
     jl_lambda_info_t *nf;
+    jl_value_t *ambiguities = jl_list_call_ambiguities_(tt, m);
+    cache = cache & (ambiguities == jl_nothing);
     if (!cache)
-        nf = jl_get_specialized(m->func.method, sig, env);
-    else
-        nf = cache_method(mt, &mt->cache, (jl_value_t*)mt, sig, tt, m, env);
+        nf = jl_get_specialized(m, sig, env);
+    else {
+        /* if (ambiguities != jl_nothing) */
+        /*     nf = NULL; */
+        /* else */
```

Comment:
I wasn't certain whether `cache=1` means "please try to insert into the cache," or whether it means "you darn well better insert into the cache" (in which case returning `nf=NULL` seems like the better option).